constraint/pkg/client/constraint_client.go

```diff
@@ -16,24 +17,58 @@ type constraintClient struct {
 	// constraint is a copy of the original Constraint added to Client.
 	constraint *unstructured.Unstructured
 
+	// matchers are the per-target Matchers for this Constraint.
+	matchers map[string]constraints.Matcher
+
 	// enforcementAction is what should be done if the Constraint is violated or
 	// fails to run on a review.
 	enforcementAction string
 
-	// matchers are the per-target Matchers for this Constraint.
-	matchers map[string]constraints.Matcher
+	// enforcementActionsForEP stores precompiled enforcement actions for each enforcement point.
+	enforcementActionsForEP map[string][]string
 }
 
 func (c *constraintClient) getConstraint() *unstructured.Unstructured {
 	return c.constraint.DeepCopy()
 }
 
-func (c *constraintClient) matches(target string, review interface{}) *constraintMatchResult {
+func (c *constraintClient) matches(target string, review interface{}, sourceEPs ...string) *constraintMatchResult {
 	matcher, found := c.matchers[target]
 	if !found {
 		return nil
 	}
 
+	enforcementActions := make(map[string]bool)
+	if apiconstraints.IsEnforcementActionScoped(c.enforcementAction) {
+		// Initialize a map to track unique enforcement actions
+		// Iterate over the provided source enforcement points (EPs)
+		for _, ep := range sourceEPs {
+			var actions []string
+			// Check if there are predefined actions for the current EP
+			if acts, found := c.enforcementActionsForEP[ep]; found {
+				actions = acts // Use the predefined actions if found
+			} else if ep == "*" {
```

Comment:
should this be checked first and exit early if it's `*` since actions is already set.

Reply:
the current implementation will override actions if another ep is evaluated after *

Reply:
fixed it.